extension/extensionlimiter/README.md

```diff
@@ -0,0 +1,305 @@
+
+Each kind of limiter as well as the wrapper type have corresponding
+**provider** interfaces that return a limiter instance based on a
+weight key or keys.
+
+Weight keys describe the standard limiting dimensions. There are
+currently four standard weight keys: network bytes, request count,
+request items, and memory size.
+
+## Key Interfaces
+
+- `LimiterWrapper`: Provides a callback-based limiting interface that
+  works with both rate and resource limiters, has a `LimitCall` method,
+  plus a provider type.
+- `RateLimiter`: Applies time-based limits, has a `Limit` method,
+  plus provider type.
+- `ResourceLimiter`: Manages physical resource limits, has
+  an `Acquire` method and a corresponding `ReleaseFunc`,
+  plus a provider type.
+- `Limiter`: Any of the above, has a `MustDeny` method.
+
+### Limiter helpers
+
+The `limiterhelper` subpackage provides:
+
+- Consumer wrappers apply limits to a collector pipeline (e.g.,
+  `NewLimitedLogs` for a limiter combined with `consumer.NewLogs`)
+- Multi-limiter combinators: for simple combined limiter functionality
+- Middleware conversion utilities: convert middleware configurations
+  to limiter providers.
+
+## Recommendations
+
+For general use cases, prefer the `LimiterWrapper` interface with its
+callback-based approach because it is agnostic to the difference between
+rate and resource limiters.
+
+Use the direct `RateLimiter` or `ResourceLimiter` interfaces only in
+special cases where control flow can't be easily scoped.
+
+Middleware configuration typically automates the configuration of
+network bytes and request count weight keys relatively early in a
+pipeline.  Receivers are responsible for limiting request items and
+memory size through one of the available helpers.
+
+Processors can apply limiters for specific reasons, for example to
+apply limits in data-dependent ways.  Exporters can apply limiters for
+the same reasons, for example to apply limits in destination-dependent
+ways.
+
+### Limiter blocking and failing
+
+Limiters implementations MAY block the request or fail immediately,
+subject to internal logic. A limiter aims to avoid waste, which
+requires balancing several factors. To fail a request that has already
+been transmitted, received and parsed is sometimes more wasteful than
+waiting for a little while; on the other hand waiting for a long time
+risks wasting memory. In general, an overloaded limiter that is
+saturated SHOULD fail requests immediately.
+
+Limiter implementations SHOULD consider the context deadline when
+they block. If the deadline is likely to expire before the limit
+becomes available, they should return a standard overload signal.
+
+### Limiter saturation
+
+All limiters feature a `MustDeny` method which is made available for
+applications to test when a limit is fully saturated. This special
+limit request is defined as the equivalent of passing a zero value to
+the limiter.
+
+Limiters SHOULD treat a request for zero units of the limit as a
```

Comment:
Would it be reasonable to make this a MUST, and get rid of the MustDeny method?

Reply:
Maybe. I like the idea, but we can keep the Limiter interface. Limiters can be defined as calling the underlying limiter with 0, and frequently we have several of them and want to call them all at once. I will separate the high-level MustDeny interface from the two low-level limiter interfaces, leaving the MultiLimiter helper, I think it will be nice.

Reply:
I think in general that the suggestion by @bogdandrutu was a good one, so I went that direction.